AssertionError in FrameworkDetector when jsDependency log something in console on initialize

java.lang.AssertionError: assertion failed
      at scala.Predef$.assert(Predef.scala:165)
      at org.scalajs.sbtplugin.FrameworkDetector.detect(FrameworkDetector.scala:68)
      at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$40.apply(ScalaJSPluginInternal.scala:455)
      at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$40.apply(ScalaJSPluginInternal.scala:440)
      ...

I have asked @sjrd in gitter about this error. He recommended me check my `jsDependencies` for case of printing to console. So it turned out.

--- a/sbt-plugin-test/build.sbt
+++ b/sbt-plugin-test/build.sbt
@@ -72,7 +72,9 @@ lazy val multiTest = crossProject.
   ).
   jsSettings(baseSettings: _*).
   jsSettings(
-    name := "Multi test framework test JS"
+    name := "Multi test framework test JS",
+    // Make FrameworkDetector resilient to other output - #1572
+    jsDependencies in Test += ProvidedJS / "consoleWriter.js"
   ).
   jvmSettings(versionSettings: _*).
   jvmSettings(

--- a/sbt-plugin-test/multiTest/js/src/test/resources/consoleWriter.js
+++ b/sbt-plugin-test/multiTest/js/src/test/resources/consoleWriter.js
@@ -1,0 +1,2 @@
+// Make FrameworkDetector resilient to other output - #1572
+console.log("Breaking message")

--- a/sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/FrameworkDetector.scala
+++ b/sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/FrameworkDetector.scala
@@ -14,6 +14,8 @@ import scala.collection.mutable
 
 final class FrameworkDetector(jsEnv: JSEnv, classpath: CompleteClasspath) {
 
+  import FrameworkDetector._
+
   /**
    *  Detects which of the test frameworks in [[frameworks]] exists on
    *  the classpath.
@@ -47,23 +49,29 @@ final class FrameworkDetector(jsEnv: JSEnv, classpath: CompleteClasspath) {
         var gotOne = false;
         for (var j = 0; j < data[i].length; ++j) {
           if (frameworkExists(data[i][j])) {
-            console.log(data[i][j]);
+            console.log("$ConsoleFrameworkPrefix" + data[i][j]);
             gotOne = true;
             break;
           }
         }
-        if (!gotOne)
-          console.log(""); // print an empty line to zip afterwards
+        if (!gotOne) {
+          // print an empty line with prefix to zip afterwards
+          console.log("$ConsoleFrameworkPrefix");
+        }
       }
     """
 
     val vf = new MemVirtualJSFile("frameworkDetector.js").withContent(code)
-    val console = new FrameworkDetector.StoreConsole
+    val console = new StoreConsole
 
     val runner = jsEnv.jsRunner(classpath, vf, NullLogger, console)
     runner.run()
 
-    val results = console.buf.toList
+    // Filter jsDependencies unexpected output
+    val results = console.buf collect {
+      case s if s.startsWith(ConsoleFrameworkPrefix) =>
+        s.stripPrefix(ConsoleFrameworkPrefix)
+    }
 
     assert(results.size == frameworks.size)
 
@@ -77,4 +85,6 @@ object FrameworkDetector {
     val buf = mutable.Buffer.empty[String]
     def log(msg: Any): Unit = buf += msg.toString
   }
+
+  private val ConsoleFrameworkPrefix = "@scalajs-test-framework-detector:"
 }

